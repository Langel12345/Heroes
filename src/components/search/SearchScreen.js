import React, { useMemo } from 'react';
import { useNavigate,useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { getHeroByName } from '../../selectors/getHeroeByName';
import { HeroCard } from '../heroes/HeroCard';
import queryString  from 'query-string';
export const SearchScreen = () => {

  const navigate=useNavigate(); //Liberira para navegar

  const location =useLocation(); // Liberira para leer los query
  
  const{ q =''} =queryString.parse(location.search);// Liberira para leer los parametros enviados

   
  const [ value,handeInputChange ]= useForm({ searchText: q });

  const {searchText} =value;

  const heroesFilter= useMemo(() => getHeroByName(q) ,[ q ] );

  const handleSearch = (e) => {
      e.preventDefault();
      navigate(`?q=${searchText}`);
    
  } 

  return <>
      <h1>Search </h1>
      <div className="row">
          <div className="col-5">
            <h4>Search </h4>
            <hr />
            <form onSubmit={handleSearch}>
              <input 
                type="text" 
                className="form-control"
                placeholder="Search Heroe"
                autoComplete='off'
                name="searchText"
                value={ searchText }
                onChange={handeInputChange}
              />

              <button 
                className="btn btn-outline-primary mt-3"
                type='submit'                 
               >
                Search...
               </button>
            </form> 
          </div>

          <div className="col-7">
              <h4>Results</h4>
              <hr />
              {
                ( q === '')
                  ? <div  className="alert alert-info">Search in Heroe</div>
                  : ( heroesFilter.length === 0)
                    && <div className="alert alert-danger">No results: { q } </div>
              }

              {
                heroesFilter.map(hero =>( 
                    <HeroCard key={hero.id}
                        {...hero}
                     />
                ))
              }
          </div>
      </div>
    </>
};
