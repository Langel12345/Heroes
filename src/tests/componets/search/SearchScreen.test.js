import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

const mockNavigate=jest.fn();
jest.mock('react-router-dom', ()=>({ 
        ...jest.requireActual('react-router-dom'),
        useNavigate :() => mockNavigate,
    })
 )

describe('Pruebas en <SearchScreen />', () => {

    test('debe de mostrarse correctamente ', () => {

        const wrapper = mount (
        <MemoryRouter initialEntries={['/search']}>
            <SearchScreen />
        </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        
        expect(wrapper.find('.alert-info').text().trim() ).toBe('Search in Heroe');
    });
    
    test('debe de mostar a Batman y el input con el valor del queryString ', () => {
      
        const wrapper = mount (
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
            );
    
        expect (wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
        
    });
    
    test(' debe de moestar un error si no se encuentra el heroe', () => {
        

        const wrapper = mount (
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchScreen />
            </MemoryRouter>
            );
            expect (wrapper.find('input').prop('value')).toBe('batman123');    
            expect (wrapper.find('.alert-danger').text().trim()).toBe('No results: batman123');
    });
    
    test(' debe de llamar el navigate a la nueva pantalla ', () => {
      
            const wrapper = mount (
            <MemoryRouter initialEntries={['/search?q=superman']}>
                <SearchScreen />
            </MemoryRouter>
            );
            
            wrapper.find('input').simulate('change',{ 
                target:"searchText",
                value:'bataman'
             });

            wrapper.find('form').prop('onSubmit')({
                preventDefault:()=>{},
            });
            expect(mockNavigate).toHaveBeenCalledWith('?q=superman');

    });
    
});
