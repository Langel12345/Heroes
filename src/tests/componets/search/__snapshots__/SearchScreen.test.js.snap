// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostar a Batman y el input con el valor del queryString  1`] = `
Array [
  <h1>
    Search 
  </h1>,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search Heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-3"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <a
        className="my-card animate__animated animate__fadeIn"
        href="/hero/dc-batman"
        onClick={[Function]}
      >
        <img
          alt="Batman"
          className="img img-responsive"
          src="/heroes/assets/dc-batman.jpg"
        />
        <div
          className="profile-name"
        >
          Batman
        </div>
        <div
          className="profile-position"
        >
          Bruce Wayne
        </div>
        <div
          className="profile-overview"
        >
          <div
            className="profile-overview"
          >
            <div
              className="row"
            >
              <div
                className="col-ms-4"
              >
                <h3>
                  DC Comics
                </h3>
                <p>
                  Primera aparici√≥n: 
                  <br />
                  Detective Comics #27
                </p>
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> debe de mostrarse correctamente  1`] = `
Array [
  <h1>
    Search 
  </h1>,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search Heroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-3"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search in Heroe
      </div>
    </div>
  </div>,
]
`;
